File: tier1_foundation/merge_intervals/solutions/p056_merge_intervals.cpp
/*
 * P56: Merge Intervals (Medium)
 * https://leetcode.com/problems/merge-intervals/
 * Topics: Array, Sorting
 *
 * Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals, and return an array of the non-overlapping intervals that cover all the intervals in the input.
 *
 * Example 1:
 *     Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
 *     Output: [[1,6],[8,10],[15,18]]
 *     Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].
 *
 * Example 2:
 *     Input: intervals = [[1,4],[4,5]]
 *     Output: [[1,5]]
 *     Explanation: Intervals [1,4] and [4,5] are considered overlapping.
 *
 * Constraints:
 *     - 1 <= intervals.length <= 104
 *     - intervals[i].length == 2
 *     - 0 <= starti <= endi <= 104
 *
 * Template (python3):
 *     class Solution:
 *         def merge(self, intervals: List[List[int]]) -> List[List[int]]:
 *
 * Hint: Sort intervals by start time, then merge overlapping ones.
 */


#include "cpptest.h"
#include <algorithm>
#include <utility>
#include <vector>

static std::vector<int> solve(const std::vector<int> &flat, int target)
{
    (void)target;
    int n = (int)(flat.size() / 2);
    if (n == 0)
        return {};

    std::vector<std::pair<int, int>> iv(n);
    for (int i = 0; i < n; i++)
        iv[i] = {flat[i * 2], flat[i * 2 + 1]};
    std::sort(iv.begin(), iv.end());

    std::vector<int> res;
    res.push_back(iv[0].first);
    res.push_back(iv[0].second);

    for (int i = 1; i < n; i++) {
        if (iv[i].first <= res.back()) {
            res.back() = std::max(res.back(), iv[i].second);
        } else {
            res.push_back(iv[i].first);
            res.push_back(iv[i].second);
        }
    }

    return res;
}

int main()
{
    TestCase cases[] = {
        {"example 1",                   {1,3,2,6,8,10,15,18}, 0, {1,6,8,10,15,18}},
        {"example 2",                   {1,4,4,5},             0, {1,5}},
        {"overlapping start",           {1,4,0,4},             0, {0,4}},
        {"single interval",             {1,1},                 0, {1,1}},
        {"all overlap into one",        {1,4,2,3,3,5},         0, {1,5}},
        {"no overlaps",                 {1,2,3,4,5,6},         0, {1,2,3,4,5,6}},
        {"interval contains others",    {1,10,2,3,4,5},        0, {1,10}},
        {"adjacent intervals touching", {1,2,2,3,3,4},         0, {1,4}},
    };

    RUN_TESTS("P56: Merge Intervals", solve, cases, 8);
}

File: tier1_foundation/merge_intervals/test_runners/p056_merge_intervals.py
from src.utils.judge_base import JudgeBase, TestCase, register


class P056MergeIntervals(JudgeBase):
    name = "56. Merge Intervals"
    test_cases = [
        TestCase(
            input=[[1, 3], [2, 6], [8, 10], [15, 18]],
            expected=[[1, 6], [8, 10], [15, 18]],
            label="example 1",
        ),
        TestCase(input=[[1, 4], [4, 5]], expected=[[1, 5]], label="example 2"),
        TestCase(input=[[1, 4], [0, 4]], expected=[[0, 4]], label="overlapping start"),
        TestCase(input=[[1, 1]], expected=[[1, 1]], label="single interval"),
        TestCase(
            input=[[1, 4], [2, 3], [3, 5]],
            expected=[[1, 5]],
            label="all overlap into one",
        ),
        TestCase(
            input=[[1, 2], [3, 4], [5, 6]],
            expected=[[1, 2], [3, 4], [5, 6]],
            label="no overlaps",
        ),
        TestCase(
            input=[[1, 10], [2, 3], [4, 5]],
            expected=[[1, 10]],
            label="interval contains others",
        ),
        TestCase(
            input=[[1, 2], [2, 3], [3, 4]],
            expected=[[1, 4]],
            label="adjacent intervals touching",
        ),
    ]

    def to_stdin(self, tc_input) -> str:
        lines = [str(len(tc_input))]
        for row in tc_input:
            lines.append(" ".join(str(x) for x in row))
        return "\n".join(lines) + "\n"

    def check_stdout(self, stdout: str, tc_expected) -> bool:
        return self.parse_int_matrix(stdout) == tc_expected


register(P056MergeIntervals)
